src/SipiImage.cpp

```diff
@@ -508,68 +508,129 @@ namespace Sipi {
     /*==========================================================================*/
 
 
-    void SipiImage::removeChan(unsigned int chan) {
-        if ((nc == 1) || (chan >= nc)) {
-            std::string msg = "Cannot remove component: nc=" + std::to_string(nc) + " chan=" + std::to_string(chan);
+    void SipiImage::removeChannel(const unsigned int channel, const bool force_gray_alpha) {
+        if ((nc == 1) || (channel >= nc)) {
+            std::string msg = "Cannot remove component: nc=" + std::to_string(nc) + " chan=" + std::to_string(channel);
             throw SipiImageError(__file__, __LINE__, msg);
         }
 
-        if (!es.empty()) {
-            if (nc < 3) {
-                es.clear(); // no more alpha channel
-            } else if (nc > 3) { // it's probably an alpha channel
-                if ((nc == 4) && (photo == SEPARATED)) {  // oh no – 4 channels, but CMYK
-                    std::string msg =
-                            "Cannot remove component: nc=" + std::to_string(nc) + " chan=" + std::to_string(chan);
-                    throw SipiImageError(__file__, __LINE__, msg);
-                } else {
-                    es.erase(es.begin() + (chan - ((photo == SEPARATED) ? 4 : 3)));
-                }
-            } else {
-                std::string msg = "Cannot remove component: nc=" + std::to_string(nc) + " chan=" + std::to_string(chan);
+        const bool has_removable_extra_samples = !es.empty();
+        const bool has_two_or_less_channels = nc < 3;
+        const bool has_three_channels = nc == 3;
+
+        // Assumtion: An image with two or less channels cannot have extra samples
+        assert(has_two_or_less_channels && has_removable_extra_samples);
+
+        if (has_removable_extra_samples) {
```

Comment:
yes, the proposed inspect thingy. Unfortunately this is still not part of the language as of c++23. Anyway, my hope is that after a thorough refactoring, this horrible if-tree will not be needed anymore. It simply tries to figure out the image characteristics. Those could, in my opinion, be better modeled, so that it is clear what image I have in hand and what operations are allowed.